logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-                return convertedRootType;
-            }
-
-            TypeVariable newRootType = new TypeVariable(Identifier.Variable.of(Reference.label(rootLabel.toString())));
-            newRootType.label(rootLabel);
-            neighbours.putIfAbsent(newRootType, new HashSet<>());
-            return newRootType;
-        }
-
-        Resolvers resolvers() {
-            return resolvers;
-        }
-
-        HashMap<TypeVariable, Set<TypeVariable>> neighbours() {
-            return neighbours;
+        Traversal resolverTraversal() {
+            return resolverTraversal;
         }
 
-        private boolean hasNoInfo(TypeVariable variable) {
-            return (!variable.reference().isLabel() &&
-                    !variable.sub().isPresent() &&
-                    !variable.label().isPresent() &&
-                    variable.is().isEmpty());
+        Variable getVariable(Reference reference) {
+            return variableRegister.get(reference);
         }
 
-        private TypeVariable convert(Variable variable) {
-            if (variable.isType()) return convert(variable.asType());
-            else if (variable.isThing()) return convert(variable.asThing());
+        private void convert(Variable variable) {
+            if (variable.isType()) convert(variable.asType());
+            else if (variable.isThing()) convert(variable.asThing());
             else throw GraknException.of(ILLEGAL_STATE);
         }
 
         private TypeVariable convert(TypeVariable variable) {
-            if (resolvers.contains(variable)) return resolvers.resolver(variable);
-
-            TypeVariable resolver = resolvers.register(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-
-            if (variable.constraints().isEmpty()) return resolver;
-
-            resolver = cloneVariable(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-            return resolver;
+            if (resolverRegister.containsKey(variable.id())) return resolverRegister.get(variable.id());
+            resolverRegister.put(variable.id(), variable);
+            variableRegister.putIfAbsent(variable.reference(), variable);
+            variable.addTo(resolverTraversal);
+            return variable;
         }
 
         private TypeVariable convert(ThingVariable variable) {
-            if (resolvers.contains(variable)) return resolvers.resolver(variable);
-
-            TypeVariable resolver = resolvers.register(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
+            if (resolverRegister.containsKey(variable.id())) return resolverRegister.get(variable.id());
 
-            if (variable.constraints().isEmpty()) return resolver;
+            TypeVariable resolver = new TypeVariable(variable.reference().isAnonymous() ?
+                                                             newSystemId() : variable.id());
```

Comment:
When a `Variable` is anonymous, we need the corresponding resolver to be referrable. This is because we want results from the traversal so we can add `resovledTypes` even to anonymous variables.

Reply:
I though `resovleLabels` would sort this for us. However, there may be circumstances where more information is gained from the full `resolvedTypes`. This is will only occur when the `TypeVariable` is a `role`.

When this occurs, we want to reset the `resolvedTypes` on a `Variable` instead of just adding them. So I've added a `setResolvedTypes()` method onto `Variable`, and call that instead of `addResovledTypes()` after the resolve Traversal has been called.